src/PaymentServer/Redemption.hs

```diff
@@ -190,40 +203,74 @@ retry op =
+  -- count.
+  -> Maybe Int
+  -- ^ Nothing if the parameters are incoherent.  Otherwise, Just the number
+  -- of tokens expected in the specified group.
+tokenCountForGroup numGroups totalTokens groupNumber
+  | totalTokens < numGroups = Nothing
+  | groupNumber >= numGroups || groupNumber < 0 = Nothing
+  | otherwise = Just $ groupSize + groupSizeAdjustment
+  where
+    (groupSize, remainder) = totalTokens `divMod` numGroups
+    groupSizeAdjustment = if groupNumber < remainder then 1 else 0
+
+
 -- | Handler for redemption requests.  Use the database to try to redeem the
 -- voucher and return signatures.  Return a failure if this is not possible
 -- (eg because the voucher was already redeemed).
-redeem :: VoucherDatabase d => Issuer -> d -> Redeem -> Handler Result
-redeem issue database (Redeem voucher tokens counter) =
-  if counter < 0 || counter >= maxCounter then
-    throwError $ jsonErr err400 (CounterOutOfBounds 0 maxCounter counter)
-  else do
-    redeemResult <- liftIO . retry $ redeem
-    case redeemResult of
-      Left NotPaid -> do
-        throwError $ jsonErr err400 Unpaid
-      Left AlreadyRedeemed -> do
-        throwError $ jsonErr err400 DoubleSpend
-      Left DuplicateFingerprint -> do
-        throwError $ jsonErr err400 $ OtherFailure "fingerprint already used"
-      Left DatabaseUnavailable -> do
-        throwError $ jsonErr err500 $ OtherFailure "database temporarily unavailable"
-      Right fresh -> do
-        let result = issue tokens
-        case result of
-          Left reason -> do
-            throwError $ jsonErr err400 $ OtherFailure reason
-          Right (ChallengeBypass key signatures proof) -> do
-            let count = fromInteger . toInteger . length $ signatures
-            -- addCounter returns bool indicating whether its argument was
-            -- greater than or equal to zero or not.  We don't care.  length
-            -- is greater than or equal to zero.
-            liftIO . when fresh $ (P.addCounter signaturesIssued count >>= \_ -> return ())
-            return $ Succeeded key signatures proof
-    where
-      fingerprint = fingerprintFromTokens tokens
-      redeem :: IO (Either RedeemError Bool)
-      redeem = redeemVoucherWithCounter database voucher fingerprint counter
+redeem :: VoucherDatabase d => RedemptionConfig -> d -> Redeem -> Handler Result
+redeem (RedemptionConfig numGroups tokensPerVoucher issue) database (Redeem voucher tokens counter) =
+  if counter < 0 || counter >= numGroups then
+    throwError $ jsonErr err400 (CounterOutOfBounds 0 numGroups counter)
+  else
+    case tokenCountForGroup numGroups tokensPerVoucher (fromIntegral counter) of
+      Nothing ->
+        throwError $ jsonErr err500 $ OtherFailure "invalid redemption counter"
+      Just allowedTokenCount ->
+        if allowedTokenCount /= length tokens
+        then throwError $ jsonErr err400 $ OtherFailure "wrong number of tokens"
```

Comment:
Should this have a structured error message, with the expected number of tokens?

Reply:
Maybe.  I guess that would be easier to answer if we had a UX for handling any of these failure conditions.  Right now what happens is ZKAPAuthorizer notes the error and usually tries exactly the same thing again a little later and no software exists to surface these error conditions to the user.